### ISSUE TITLE
Pylance infers Literal[] type where it is not intended

### ISSUE THREAD

**Issue by tthommes** [microsoft/pylance-release#4633]
When I try to pass an Enum-value e.g.
```
class MyEnum(Enum):
  A = 0x1
  B = 0x2
```
to a function parameter type-hinted as `Integral` e.g.
```
from numbers import Integral
def myfunc(a: Integral):
  pass

myfunc(MyEnum.A.value)
```
pylance infers `Literal[]` for the Enum-value and complains about incompatibility.

_Originally posted by @tthommes in https://github.com/microsoft/pylance-release/issues/697#issuecomment-1643939938_

There has been a bug, fixed three years ago, that seems to be the same issue. When I back-port my Pylance-Version to that one, the here described issue is gone. So I suppose the bug was somehow "resurrected".

**Comment by erictraut** [comment]
This behavior is correct from a type checking perspective. An `int` value is not type compatible with the class `Integral`.

```python
x: Integral = 0 # Type error
```

Not surprisingly, mypy generates the same error in this case.

The abstract numeric types defined in `numbers` have never been integrated into the static type system for Python. There has been much discussion about how this might be done, but no serious proposals have been put forth by the typing community.

The recommended workaround is to avoid the use of abstract numeric types and use the standard builtin types like `int`.

**Comment by erictraut** [comment]
Another workaround, if you really want to use `Integral`, is to define your own custom type that derives from both `int` and `Integral` and use that in your enum definition.

```python
class MyInt(int, Integral):
    pass


class MyEnum(Enum):
    A = MyInt(0x1)
    B = MyInt(0x2)
```

**Comment by tthommes** [comment]
thanks for the quick response. The first and recommended workaround works fine.
With the second workaround I however get "Cannot instanciate abstract class `MyInt`" which is probably due to the shortness of the type-definition...

**Comment by erictraut** [comment]
Are you receiving the "Cannot instanciate abstract class" error from the Python interpreter or VS Code? It runs fine for me. Here's the full sample:

```python
from enum import Enum
from numbers import Integral

class MyInt(int, Integral):
    pass

class MyEnum(Enum):
    A = MyInt(0x1)
    B = MyInt(0x2)

def myfunc(a: Integral):
    pass

myfunc(MyEnum.A.value)
```

> which however is probably not an issue of Pylance but rather the `numbers` module?

The types in the `numbers` module predate static typing in Python. The definitions for these types have never been updated to work with the static type system. They should probably be changed from abstract (nominal) classes to protocols (structural types). Because they're defined as nominal, they must appear within the class hierarchy of a given class for a type checker to consider them "type compatible". Since `int` doesn't have `Integral` in its class hierarchy, a static type checker will necessarily see the two as incompatible by the rules of the Python static type system.

**Comment by tthommes** [comment]
The abstract class error is reported by Pylance in VS Code:
```
Cannot instantiate abstract class "MyInt"
  "Number.__hash__" is abstract Pylance(reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
> The abstract class error is reported by Pylance in VS Code

Looks like you're using strict-mode type checking with the latest production version of Pylance. If you use the latest prerelease version of Pylance, this issue doesn't occur. That's because the prerelease version of Pylance contains a newer version of the typeshed stubs, and a small change was made recently in these stubs to include `__hash__` in the definition of the `int` class.

**Comment by rchiodo** [comment]
Closing as this is fixed in prerelease version 2023.7.31.
